util/crc32c.cc

```diff
@@ -475,11 +468,566 @@ std::string IsFastCrc32Supported() {
+            case 7:
+              CRCtriplet(crc, next, -7);
+            case 6:
+              CRCtriplet(crc, next, -6);
+            case 5:
+              CRCtriplet(crc, next, -5);
+            case 4:
+              CRCtriplet(crc, next, -4);
+            case 3:
+              CRCtriplet(crc, next, -3);
+            case 2:
+              CRCtriplet(crc, next, -2);
+            case 1:
+              CRCduplet(crc, next, -1); // the final triplet is actually only 2
+              //{ CombineCRC(); }
+              crc0 = CombineCRC(block_size, crc0, crc1, crc2, next2);
+              if (--n > 0) {
+                crc1 = crc2 = 0;
+                block_size = 128;
+                // points to the first byte of the next block
+                next0 = next2 + 128;
+                next1 = next0 + 128; // from here on all blocks are 128 long
+                next2 = next1 + 128;
+              }
+            case 0:;
+          } while (n > 0);
+      }
+      next = (const unsigned char*)next2;
+    }
+    uint64_t count2 = len >> 3; // 216 of less bytes is 27 or less singlets
+    len = len & 7;
+    next += (count2 * 8);
+    switch (count2) {
+      case 27:
+        CRCsinglet(crc0, next, -27 * 8);
+      case 26:
+        CRCsinglet(crc0, next, -26 * 8);
+      case 25:
+        CRCsinglet(crc0, next, -25 * 8);
+      case 24:
+        CRCsinglet(crc0, next, -24 * 8);
+      case 23:
+        CRCsinglet(crc0, next, -23 * 8);
+      case 22:
+        CRCsinglet(crc0, next, -22 * 8);
+      case 21:
+        CRCsinglet(crc0, next, -21 * 8);
+      case 20:
+        CRCsinglet(crc0, next, -20 * 8);
+      case 19:
+        CRCsinglet(crc0, next, -19 * 8);
+      case 18:
+        CRCsinglet(crc0, next, -18 * 8);
+      case 17:
+        CRCsinglet(crc0, next, -17 * 8);
+      case 16:
+        CRCsinglet(crc0, next, -16 * 8);
+      case 15:
+        CRCsinglet(crc0, next, -15 * 8);
+      case 14:
+        CRCsinglet(crc0, next, -14 * 8);
+      case 13:
+        CRCsinglet(crc0, next, -13 * 8);
+      case 12:
+        CRCsinglet(crc0, next, -12 * 8);
+      case 11:
+        CRCsinglet(crc0, next, -11 * 8);
+      case 10:
+        CRCsinglet(crc0, next, -10 * 8);
+      case 9:
+        CRCsinglet(crc0, next, -9 * 8);
+      case 8:
+        CRCsinglet(crc0, next, -8 * 8);
+      case 7:
+        CRCsinglet(crc0, next, -7 * 8);
+      case 6:
+        CRCsinglet(crc0, next, -6 * 8);
+      case 5:
+        CRCsinglet(crc0, next, -5 * 8);
+      case 4:
+        CRCsinglet(crc0, next, -4 * 8);
+      case 3:
+        CRCsinglet(crc0, next, -3 * 8);
+      case 2:
+        CRCsinglet(crc0, next, -2 * 8);
+      case 1:
+        CRCsinglet(crc0, next, -1 * 8);
+      case 0:;
+    }
+  }
+  {
+    align_to_8(len, crc0, next);
+    return (uint32_t)crc0;
+  }
+}
+
+#endif //HAVE_SSE42
+
+static inline Function Choose_Extend() {
+#ifndef HAVE_POWER8
+  if (isSSE42()) {
+#ifndef NO_THREEWAY_CRC32C
+    return crc32c_3way;
```

Comment:
IMO, it'd be good to do a dynamic check for pclmul features here and fall back to the old fast crc32c otherwise. Nehalem is one architecture that supports sse4.2 but not pclmul.

Reply:
Sure. but if we use the compile test in build_detect_platform and set HAVE_SSE42 (or whatever name), this test is no longer necessary, right?

Reply:
one is to test whether the compiler has it, the other is to test whether the machine it's running on has it. I think that's the same reason why we have isSSE42() even though HAVE_SSE42 is already set.